Mathlib/CategoryTheory/Limits/FormalCoproducts.lean

```diff
@@ -0,0 +1,434 @@
+@[simps!] def evalOf : eval C A ⋙ (whiskeringLeft _ _ A).obj (of C) ≅ Functor.id (C ⥤ A) :=
+  NatIso.ofComponents fun F ↦ NatIso.ofComponents
+    (fun x ↦ ⟨Sigma.desc fun _ ↦ 𝟙 _, Sigma.ι (fun _ ↦ F.obj x) PUnit.unit, by aesop, by simp⟩)
+    (fun f ↦ Sigma.hom_ext _ _ (by simp [Sigma.ι_desc]))
+
+variable {C A}
+
+/-- `eval(F)` preserves arbitrary coproducts. -/
+def isColimitEvalMapCocone : IsColimit (((eval.{w} C A).obj F).mapCocone (cofan.{w} J f)) where
+  desc s := Sigma.desc fun i ↦ Sigma.ι (F.obj ∘ (f i.1).obj) i.2 ≫ s.ι.app ⟨i.1⟩
+  fac s i := Sigma.hom_ext _ _ fun i ↦ by simp [cofan, Function.comp_def]
+  uniq s m h := Sigma.hom_ext _ _ fun ⟨i₁, i₂⟩ ↦ by simp [← h, cofan, Function.comp_def]
+
+instance : PreservesColimit (Discrete.functor f) ((eval.{w} C A).obj F) :=
+  ⟨fun hc ↦ ⟨IsColimit.ofIsoColimit (isColimitEvalMapCocone J f F)
+    ((Cocones.functoriality _ _).mapIso ((isColimitCofan J f).uniqueUpToIso hc))⟩⟩
+
+instance : PreservesColimitsOfShape (Discrete J) ((eval.{w} C A).obj F) :=
+  preservesColimitsOfShape_of_discrete _
+
+end HasCoproducts
+
+
+noncomputable section HasProducts
+
+variable [HasProducts.{w} A] (C) (J : Type w) (f : J → FormalCoproduct.{w} C) (F : Cᵒᵖ ⥤ A)
+
+/-- A presheaf valued in a category `A` with arbitrary products can be extended to the category of
+formal coproducts. -/
+@[simps!] def evalOp : (Cᵒᵖ ⥤ A) ⥤ ((FormalCoproduct.{w} C)ᵒᵖ ⥤ A) where
+  obj F :=
+    { obj X := ∏ᶜ fun (i : X.unop.I) ↦ F.obj (op (X.unop.obj i))
+      map f := Pi.lift fun i ↦ Pi.π _ (f.unop.f i) ≫ F.map (f.unop.φ i).op }
+  map α := { app f := Pi.map fun i ↦ α.app (op (f.unop.obj i)) }
+
+/-- `evalOp(F)` restricted to the original category (via `of`) is the original presheaf `F`. -/
+@[simps!] def evalOpOf :
+    evalOp C A ⋙ (whiskeringLeft _ _ A).obj (of C).op ≅ Functor.id (Cᵒᵖ ⥤ A) :=
+  NatIso.ofComponents fun F ↦ NatIso.ofComponents fun x ↦
+    ⟨Pi.π _ PUnit.unit, Pi.lift fun _ ↦ 𝟙 _, by aesop, by simp⟩
+
+variable {C A}
+
+/-- `evalOp(F)` preserves arbitrary products. -/
+def isLimitEvalMapCone : IsLimit (((evalOp.{w} C A).obj F).mapCone (cofan.{w} J f).op) where
+  lift s := Pi.lift fun i ↦ s.π.app ⟨i.1⟩ ≫ Pi.π _ i.2
+  fac s i := Pi.hom_ext _ _ fun i ↦ by simp [Pi.lift_π, cofan]
+  uniq s m h := Pi.hom_ext _ _ fun ⟨i₁, i₂⟩ ↦ by simp [Pi.lift_π, ← h, cofan]
+
+instance : PreservesLimit (Discrete.functor (op ∘ f)) ((evalOp.{w} C A).obj F) :=
+  ⟨fun hc ↦ ⟨IsLimit.ofIsoLimit (isLimitEvalMapCone J f F) ((Cones.functoriality _ _).mapIso
+    ((Cofan.IsColimit.op (isColimitCofan J f)).uniqueUpToIso hc))⟩⟩
+
+end HasProducts
+
+
+/-- A family of maps with the same target can be turned into one arrow in the category of formal
+coproducts. This is used in Čech cohomology. -/
+@[simps!] def arrowOfMaps (X : C) {J : Type w} (f : (j : J) → C) (φ : (j : J) → f j ⟶ X) :
+    FormalCoproduct.mk _ f ⟶ (of C).obj X :=
+  ⟨fun _ ↦ PUnit.unit, φ⟩
```

Comment:
Perhaps this construction could be moved earlier in the file for readability? It seems like it can be next to other constructors for morphisms, e.g maybe after `ofHomEquiv`?

I’m also not sure what the right name for it should be, but I think in Mathlib morphisms are either called `hom` rather than "arrow" or "maps", so I’m not sure the current name is a good fit.

Since it can be thought of as a `IsColimit.desc` seen through the isomorphism `(cofan X.I X.toFun).pt ≅ X`, maybe it could be called `ofDesc`(or `inclDesc` if `of` gets renamed?), but I’m not super convinced by this suggestion myself.

Reply:
I don’t think Cech cohomology belong in this file, so there is no reason to order things in this file so that they "makes sense" for some other file.

Reply:
done